sstables/sstable_set.cc

```diff
@@ -946,9 +946,10 @@ compound_sstable_set::compound_sstable_set(schema_ptr schema, std::vector<lw_sha
 std::unique_ptr<sstable_set_impl> compound_sstable_set::clone() const {
     std::vector<lw_shared_ptr<sstable_set>> cloned_sets;
     cloned_sets.reserve(_sets.size());
-    for (auto& set : _sets) {
+    for (const auto& set : _sets) {
         // implicit clone by using sstable_set's copy ctor.
-        cloned_sets.push_back(make_lw_shared(std::move(*set)));
+        auto cloned_set = *set;
+        cloned_sets.push_back(cloned_set.shared_from_this());
```

Comment:
I don't understand what is happening here. Who keeps `cloned_set` alive? We seem to allocate it on the stack and then create a shared pointer from it -- but the shared pointer doesn't own the storage and it will call `free()` on a stack pointer.

Reply:
oops you're right.
the previous version (calling `make_lw_shared(*set)`) was confusing but apparently correct.

Reply:
How did the tests not catch this? This seems like a quite fundamental operation, the debug tests should have caught both the use-after-fee and the attempted free of a non-heap pointer.

Reply:
> How did the tests not catch this? This seems like a quite fundamental operation, the debug tests should have caught both the use-after-fee and the attempted free of a non-heap pointer.

Worth verifying.

Reply:
> I don't understand what is happening here. Who keeps `cloned_set` alive? We seem to allocate it on the stack and then create a shared pointer from it -- but the shared pointer doesn't own the storage and it will call `free()` on a stack pointer.

Good catch. Missed this completely.

Reply:
Will make @bhalevy happy and patch it like this:
```
-        auto cloned_set = *set;
-        cloned_sets.push_back(cloned_set.shared_from_this());
+        auto cloned_set = make_lw_shared(*set);
+        cloned_sets.push_back(std::move(cloned_set));

```

Reply:
Looks like a step back for me from `cloned_sets.push_back(make_lw_shared(*set))`, but fine.
Please look into why the tests didn't catch the previous broken code.

Reply:
> Looks like a step back for me from `cloned_sets.push_back(make_lw_shared(*set))`, but fine. Please look into why the tests didn't catch the previous broken code.

I documented the behavior above the statement:
"// implicit clone by using sstable_set's copy ctor."

So probably the additional row for making it clearer to reader that set was closed isn't really needed. @bhalevy thoughts?

Reply:
> Please look into why the tests didn't catch the previous broken code.

Looking now

Reply:
we were really lacking coverage in unit test. 

A simple snipped like this in a compound unit test will trigger ASAN

+        {
+            auto cloned_compound = *compound;
+            BOOST_REQUIRE(cloned_compound.all()->size() == 3);
+        }

Reply:
> So probably the additional row for making it clearer to reader that set was closed isn't really needed. @bhalevy thoughts?

I'm fine with the extra line. Personally I preferred the single-line variant but we already spent way too much time discussing the color of this bike-shed.